Validate storage texture formats when creating BGLs

andyleiserson opened the pull request:
We were validating storage texture compatibility when creating bindings, not when creating BGLs, which is when the spec calls for it. Move the existing validation to BGL creation. binding creation already checked that the format matches the layout, so we don't need to re-validate the format when the binding is created.

I omitted a changelog entry since it seems unlikely that anybody could be accomplishing anything useful by creating such BGLs and not binding them, but, I'd be happy to add one if it seems appropriate. (Not specific to this change, but in general, stricter validation seems likely to break things and thus worth noting in the change log. On the other hand, we have a lot of missing validation so noting absolutely every change to validation could bloat the changelog quite a bit.)

**Testing**
By the CTS.

**Squash or Rebase?** Squash

**Checklist**

- [ ] If this contains user-facing changes, add a `CHANGELOG.md` entry.

andyleiserson commented on wgpu-core/src/device/resource.rs:
```diff
@@ -3376,6 +3391,47 @@ impl Device {
 
+    /// Validates that storage texture format features support the requested access mode
+    /// and returns the corresponding internal TextureUses flags.
+    fn validate_storage_texture_format_support(
```
I think it probably makes sense to move this helper to a free function in `conv` rather than being an associated method on `Device`. The reason I haven't is that I briefly considered whether the call to `describe_format_features` should be inside the helper, in which case a method on `Device` would be needed. So I thought I would wait and see if there is review feedback.

andyleiserson commented on wgpu-core/src/binding_model.rs:
```diff
@@ -59,6 +59,8 @@ pub enum BindGroupLayoutEntryError {
     #[error(transparent)]
     Device(#[from] DeviceError),
+    #[error(transparent)]
+    Binding(#[from] BindingError),
```
I added this here to avoid `match`-conversion of errors returned by the helper or introducing a new error type. But I can see a case that it makes more sense to preserve `BindingError` exclusively for bindings and not binding templates.

teoxoy reviewed (commented):
The spec doesn't seem to require validating this twice. The format is validated when creating the BGL and when creating the BG, the format of the view is checked against the format of the BGL entry.

andyleiserson committed:
fix(core): Validate storage texture formats when creating BGLs

This moves existing validation from binding creation to BGL
creation. Binding creation already had a check that returns
`CreateBindGroupError::InvalidStorageTextureFormat` if the format
does not match the layout, so it is not necessary to revalidate
the format support at that point.

andyleiserson commented:
Updated to validate format support only on BGL creation. The existing check in binding creation that the format matches the layout remains.
